internals/network/concurrent_downloader.go

```diff
@@ -20,44 +20,45 @@ func acceptedStatusCodes(code int) bool {
 }
 
 // ConcurrentDownloader: Concurrently download the resource with specified concurrency value.
-func ConcurrentDownloader(meta *FileMeta, thread int, outputName string) {
-	fmt.Println("Download the file in threads: ", thread)
+// it returns download status as bool
+func ConcurrentDownloader(meta *FileMeta, thread int, sessionID string) (*internals.Chunks, bool) {
+	isDownloadComplete := true
 	chunks := internals.Chunks{Count: thread, TotalSize: int(meta.ContentLength)}
 	chunks.ComputeChunks()
 	var wg sync.WaitGroup
 	for i, segment := range chunks.Segments {
+
 		// if segment exist skip current segment download.
-		if pkg.FileExists(internals.SegmentFilePath(internals.SessionID, i)) {
+		if pkg.FileExists(internals.SegmentFilePath(sessionID, i)) {
 			// fmt.Println("Segment Id: ", i, "already downloaded").
 			continue
 		}
 		request, err := BuildRequest(http.MethodGet, meta.FileURL)
 		if err != nil {
-			fmt.Println(err)
+			fmt.Println("Could not build request: ", err)
+			isDownloadComplete = false
 		}
 		wg.Add(1)
+
 		i := i
 		segment := segment
+
+		// Downoading individual segments in go routines
 		go func() {
 			defer wg.Done()
-			err = DownloadSegment(request, i, segment)
+			err = DownloadSegment(request, i, segment, sessionID)
 			if err != nil {
+				isDownloadComplete = false
```

Comment:
Here, why are we sharing a resource(isDownloadComplete) among multiple goroutines?
This is a clear invitation for the RACE condition.

see this:  https://stackoverflow.com/questions/39207608/how-does-golang-share-variables-between-goroutines

Reply:
Right, let's figure out a better way to achieve the same.